core/src/main/java/org/pentaho/di/connections/utils/ConnectionUriParser.java

```diff
@@ -0,0 +1,229 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details.
+ *
+ * Copyright (c) 2024 Hitachi Vantara. All rights reserved.
+ */
+
+package org.pentaho.di.connections.utils;
+
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Helper class to extract information from the
+ * {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI and more general VFS URIs.
+ * This class handles special characters in the connection name section of the
+ * {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI by simply parsing
+ * the connection name as is. This class has limited functionality.
+ * <p/>
+ * Main motivation for this class is not being able to use directly {@link java.net.URI}
+ * due to special characters in the {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI.
+ * Future efforts should look into {@link org.apache.commons.vfs2.provider.UriParser#encode(String, char[])}.
+ *
+ * @see #SPECIAL_CHARACTERS_FULL_SET
+ */
+public class ConnectionUriParser {
+
+  /**
+   * Full set of special characters the connection name can be. Only excluding the character '/' which is a file path deliminator.
+   * <p/>
+   * Special characters in this context are characters that generally have to be encoded
+   * otherwise the use of {@link java.net.URI } will throw {@link java.net.URISyntaxException} when during
+   * object instantiation or subsequent method calls.
+   *
+   */
+  public static final String SPECIAL_CHARACTERS_FULL_SET = "!\"#$%&'()*+,-.:;<=>?@[\\] \t\n\r^_`{|}~";
+
+  /**
+   * Full set of alphanumeric characters the connection name can be.
+   */
+  public static final String ALPHANUMERIC_CHARACTERS_FULL_SET =
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+  /**
+   * Full set of accepted characters the connection name can be.
+   */
+  public static final String ACCEPTED_CHARACTERS_FULL_SET =
+      SPECIAL_CHARACTERS_FULL_SET + ALPHANUMERIC_CHARACTERS_FULL_SET;
+
+  /**
+   * Pattern to match a {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI
+   *  with scheme, connection name, and connection path.
+   * <p/> Some Examples:
+   * <ul>
+   *   <li>For PVFS URI: "pvfs://connectionName/someFolderA/SomeFolderB/someFile.txt"</li>
+   *      <ul>
+   *        <li>scehme: "pvfs"</li>
+   *        <li>connection Name: "connectionName"</li>
+   *        <li>pvfs path: "/someFolderA/SomeFolderB/someFile.txt"</li>
+   *
+   *      </ul>
+   * </ul>
+   * Regex should be encompass {@link org.pentaho.di.connections.vfs.provider.ConnectionFileSystem#DOMAIN_ROOT}
+   */
+  public static final Pattern CONNECTION_URI_WITH_CONNECTION_NAME_AND_PATH_PATTERN
+    = Pattern.compile(  "(\\w+)://([^/]+)(/.+)" );
```

Comment:
Shouldn't these patterns be anchored at the beginning and end?

Reply:
I would say no. It doesn't matter in this case. But I can add it.

with the current code, I test 20+ strings for the various format and behave as expected.
And adding anchor, `'^'`, when I retested, still 20+ strings pass.

I read on this article https://stackoverflow.com/questions/1894624/is-regex-in-java-anchored-by-default-with-both-a-and-character

- I'm using Mather#find which isn't anchored with ^ and $.

the code now works with space before a URI so "     pvfs://blah/blah". I couldn't see an event where any Pentaho would be code is passing around URI with padding before.

The only advantage would be to if `ConnectionUriParser` was passed a bad URI such as "     pvfs://blah/blah", then

-  the parsing would fail gracefully, not throw an exception
- and all the getters like `getScheme()` would return null.

Reply:
Addressed in commit: https://github.com/pentaho/pentaho-kettle/commit/57a7b9ed55209fb93c39549b26b39f2a50442921

Reply:
Thanks! I was thinking more of weird cases such as `/foo/scheme://bucket/folder\\trailer`.